### ISSUE TITLE
The effect library supports Display-P3 color primaries

### ISSUE THREAD

**Issue by GavinChu** [androidx/media#1915]
When using Display P3 video through the video frame processor, the displayed image will be white. Currently, the effect library does not support DisPlay-P3.

**DisPlay-P3 video color parameters:**
colour_range: Full
colour_primaries: Display P3
transfer_characteristics: sRGB/sYCC
matrix_coefficients: BT.601
matrix_coefficients_Original: BT.470 System B/G

**Comment by claincly** [comment]
Do you mean the [DCI-P3](en.wikipedia.org/wiki/DCI-P3) color? What confuses me is the combination of P3/sRGB and BT.601. Regardless, the output should not be completely white. Do you have a test media we can use to reproduce?

**Comment by GavinChu** [comment]
Oh, I'm very sorry, there are some problems with my description. After using the effect processing, the displayed color is a bit pale, not the real visual color. The P3 D65 color is used from the file parsing, and the value of color primaries is 12. In addition, I think both P3 D65 and DCI P3 need to support effect processing.

**Comment by GavinChu** [comment]
P3-D65 Test Media:
[test_video_P3-D65.zip](https://github.com/user-attachments/files/17911863/test_video_P3-D65.zip)

**Comment by claincly** [comment]
Thanks for the media.

We don't support extensive color space management yet, so you might need to hack the library a bit to get correct color - I'll try to find a way to do it.

OOI, I wonder how is the file generated? It seems like a video recording from a phone, but why is it in P3 and using full range color? My understanding is P3 is usually for displays and recorded media usually use limited range color.

**Comment by GavinChu** [comment]
Yes, the video was recorded by a mobile phone. The videos currently recorded by iOS may have P3 colors. We need to support the playback of this color, although there are not many such videos.

Using SRGB_BT709_FULL did not bring any effect changes, and I think it seems to be useless. FULL and LIMITED are not actually processed for SDR videos in glsl.

I tried to modify the code to support the recognition of P3, in which MediaCodec does not support the setting of P3 color parameters. And MediaCodec inputs RGB data of P3 color to GL by default.

We have to do some processing for P3, and I tried two solutions:

1. EGL windowAttributes is set to EXT_gl_colorspace_display_p3_passthrough, which is very different from directly playing P3 (without effect processing). Another point I haven't figured out why EGL_EXT_gl_colorspace_display_p3 and EGL_EXT_gl_colorspace_display_p3_linear have no effect.
2. I referred to the conversion from BT2020 to BT709 in *external_yuv_es3.glsl, and implemented the conversion from P3 D65 to BT709 in *sdr_external_es2.glsl, and the effect is basically the same as EXT_gl_colorspace_display_p3_passthrough.

For the second solution, there is an error between you and me in the calculation of XYZ_TO_RGB_BT709. My calculation result is {3.24100423f, -0.96922427f, 0.05563935f, -1.53739941f, 1.87592983f, -0.20401104f, -0.49861604f, 0.04155425f, 1.05714846f}.
I used the BT709 parameters in android.graphics.ColorSpace and the computeXYZMatrix function for calculation (similarly, the calculation of RGB_TO_XYZ_BT2020 also has errors). I am not sure whether these errors will affect the effect.

The effect after conversion is very different from directly playing P3 (without effect processing). I suspect that it is caused by the incorrect use of 601 matrix conversion in MediaCodec, which has not been verified yet.

**Comment by claincly** [comment]
Which player are you using? ExoPlayer or CompositionPlayer?

The combination of P3 and BT.601 is also strange

**Comment by GavinChu** [comment]
I use ExoPlayer + setVideoEffects

**Comment by claincly** [comment]
Thanks. We tried internally and it seems the color is a tiny bit different with our unmodified pipeline. So for your case, 

- If `EXT_gl_colorspace_display_p3_passthrough` works for you, that's great, but there's not an easy way of using it without forking the library. So use it with caution
- The second way, by defining your own color transfer matrix, will be more scalable. I would recommend taking a look at [`RgbMatrix`](https://github.com/androidx/media/blob/release/libraries/effect/src/main/java/androidx/media3/effect/RgbMatrix.java), work out your own transformation matrix and plug it into `setVideoEffects()`

**Comment by GavinChu** [comment]
OK, thanks, I'll try it
